### ISSUE TITLE
Show the currently focused cell by other collaborators

### ISSUE THREAD

**Issue by martinRenou** [jupyterlab/jupyterlab#11555]
Similar to https://github.com/jupyterlab/jupyterlab/issues/11528 

### Problem

We currently have a way to see the cursor of other collaborators. But this only works when the other collaborators are in "edit" mode. This results in this issue https://github.com/jupyterlab/jupyterlab/issues/11406

We should maybe think of a way to indicate that another collaborator has focus on one cell, whether they are in "edit" mode or "command" mode. Similar to what google does for showing the focused cells in spreadsheets:
![spreadsheet](https://user-images.githubusercontent.com/21197331/143556116-264748ec-aaad-4434-83b6-cddbc75d3c36.png)

### Proposed Solution

Show an outline colored with the collaborator's color on the focused cell.

### Additional context

This would fix https://github.com/jupyterlab/jupyterlab/issues/11406

**Comment by hbcarlos** [comment]
Maybe we can show a list of collaborators in each cell to see where is everyone.
Something like this, but with smaller icons.

![rendered-markdown](https://user-images.githubusercontent.com/26092748/143570264-17c084ed-35d2-4139-8b62-320d277cdd53.png)

**Comment by martinRenou** [comment]
It's similar to what I wanted to suggest in https://github.com/jupyterlab/jupyterlab/issues/11528#issuecomment-976651851. Though if we already have the list of contributors somewhere else in the UI it might be too much redundant information.

Another idea could be to just have a similar visual indication to the "active" state:
![collab](https://user-images.githubusercontent.com/21197331/143571505-983e643c-1a24-4eff-817f-893a974b3eee.png)

**Comment by krassowski** [comment]
We could even show the cursor. It is technically feasible - while figuring out where the cursor should be after rendering is difficult, the trick is to insert the cursor before rendering so that it is reserved by marked.js.

We do this already for rendering Tex (we use custom markers to substitute stuff after rendering).

I think that rendering the cursor would be the best UX possible.

Out of the two proposals above I slightly prefer the icons over colours (you know colourblind people etc).

**Comment by martinRenou** [comment]
> I think that rendering the cursor would be the best UX possible. (of course this is for edit mode, for command mode the icons or colors are fine).

We're doing this already in edit mode. Or did I misunderstand what you mean by cursor?

**Comment by krassowski** [comment]
I mean when person A is in edit mode but person B has markdown rendered, we are currently not showing the cursor of A in the rendered markdown of person B.

Instead PR #11544 attempts to switch person B to unrendered state which leads to UX issues discussed on that PR (unable to have rendered preview in second window, unable to render while collaborator is in the same cell)  and more (potential performance penalty from switching between states).

Apologies, maybe I should have posted this as a comment on that PR. I misinterpreted this as a discussion revisiting that topic, while it mostly focuses on the command mode.

**Comment by martinRenou** [comment]
Yes, thank you for the clarification. Indeed both showing the cursor in the rendered Markdown cells in edit mode and showing the focused cell in command mode could be complementary.

**Comment by martinRenou** [comment]
I pushed a POC on [this branch](https://github.com/martinRenou/jupyterlab/tree/jp-user) which is on top of @hbcarlos IUser changes.

https://user-images.githubusercontent.com/21197331/143599662-05919493-dbe4-4456-aba9-c2ae2afd63b9.mp4

I personally like the result, but it goes in the way of the dirty cell state visual representation. Also how do we show that there are more than one contributor?

On another note, it looks like [deepnote shows a list of collaborator icons next to the cell](https://deepnote.com/static/landing-new/video.mp4), as suggested by Carlos.

**Comment by jasongrout** [comment]
> On another note, it looks like [deepnote shows a list of collaborator icons next to the cell](https://deepnote.com/static/landing-new/video.mp4), as suggested by Carlos.

I like the idea of user icons next to the cell instead of a frame around the cell. I think the icon gives more info (i.e., who, not just a color), and having a frame around the cell is I think to heavy of a change UX-wise.

**Comment by martinRenou** [comment]
I've been recently working on the following approach:

![collabicons](https://user-images.githubusercontent.com/21197331/145253806-75559660-e98c-43e4-a0c5-4b7a46b8ccc9.png)


See this branch: https://github.com/martinRenou/jupyterlab/tree/collaborator_list (not opening a PR yet, as it's depending on Carlos's IUser PR)

**Comment by martinRenou** [comment]
As discussed during the weekly meeting, there are some other options we could explore:

- put the icons on the line gutter (right or left) and remove the flag on the other collaborator cursors.
  - Pros: This would add the information of the line number people are working on. 
  - Cons: It will not work well with rendered markdown cells.
- put the icon list outside of the cell.
  - Pros: It prevents the icons to hide the code.
  - Cons: It will take more space in the Notebook UI than the solution proposed in the above comment.

**Comment by davidbrochart** [comment]
What about putting the icons inside the cell but making them transparent, so that they appear "behind" the code?

**Comment by fcollonval** [comment]
Moving this to jupyter-collaboration
